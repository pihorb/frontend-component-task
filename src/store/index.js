export {context} from './context';
export {reducer} from './reducer';
export {types} from './types';
