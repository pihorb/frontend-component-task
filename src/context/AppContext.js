import React, {createContext, useContext, useReducer} from 'react';
import reducer from '../reducers/reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from '../reducers/types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const setCheckBoxes = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    checkBoxes: {
      gems: setCheckBoxes('create', 'update'),
      folders: setCheckBoxes('create', 'update')
    }
  });

  const setActiveRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const updateCheckBoxes = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggle: setActiveRole,
        updateCheckBoxes,
        role: state.role,
        checkBoxes: state.checkBoxes,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
