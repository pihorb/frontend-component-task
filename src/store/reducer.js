import {
  ADD_FOLDERS,
  ADD_GEMS,
  ADMIN,
  CUSTOM,
  MEMBER,
  REMOVE_FOLDERS,
  REMOVE_GEMS,
  USER
} from './types';
import {defaultSelected} from './context';

const reducer = (state, action) => {
  switch (action.type) {
    case USER:
      return {
        ...state,
        role: 'user',
        slots: {
          gems: defaultSelected('create', 'update'),
          folders: defaultSelected('create', 'update')
        }
      };
    case MEMBER:
      return {
        ...state,
        role: 'member',
        slots: {
          gems: defaultSelected(),
          folders: defaultSelected()
        }
      };
    case ADMIN:
      return {
        ...state,
        role: 'admin',
        slots: {
          gems: defaultSelected('create', 'update', 'move', 'delete'),
          folders: defaultSelected('create', 'update', 'move', 'delete')
        }
      };
    case CUSTOM:
      return {
        ...state,
        role: 'custom',
        slots: {
          gems: [],
          folders: []
        }
      };
    case ADD_GEMS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems, action.payload],
          folders: [...state.slots.folders]
        }
      };
    case ADD_FOLDERS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems],
          folders: [...state.slots.folders, action.payload]
        }
      };
    case REMOVE_GEMS:
      return {
        ...state,
        slots: {
          gems: state.slots.gems.filter((item) => item !== action.payload),
          folders: [...state.slots.folders]
        }
      };
    case REMOVE_FOLDERS:
      return {
        ...state,
        slots: {
          gems: [...state.slots.gems],
          folders: state.slots.folders.filter((item) => item !== action.payload)
        }
      };
    default:
      return state;
  }
};

export default reducer;
