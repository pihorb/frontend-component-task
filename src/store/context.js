import React, {createContext, useContext, useReducer} from 'react';
import reducer from './reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from './types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const defaultSelected = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    slots: {
      gems: defaultSelected('create', 'update'),
      folders: defaultSelected('create', 'update')
    }
  });

  const toggleRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const toggleSelect = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggleRole,
        toggleSelect,
        role: state.role,
        slots: state.slots,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
