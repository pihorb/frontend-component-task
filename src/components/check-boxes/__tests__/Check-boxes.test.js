import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import React from 'react';
import {cleanup, fireEvent, render} from '@testing-library/react';
import CheckBoxes from '../Check-boxes';
import {AppProvider} from '../../../store/context';

const renderCheckBoxes = (cb, path) => {
  return cb(
    <AppProvider>
      <CheckBoxes path={path} />
    </AppProvider>
  );
};

beforeEach(cleanup);

describe('checkBoxes component test', () => {
  it('renders correctly', () => {
    const {queryByTestId} = renderCheckBoxes(render, 'folders');

    expect(queryByTestId('check-boxes')).toBeTruthy();
  });

  it('check class and value', () => {
    const {queryByTestId} = renderCheckBoxes(render, 'spam');
    const title = queryByTestId('check-boxes').firstChild;

    expect(title.classList.contains('check-boxes__title')).toBe(true);
    expect(title).toHaveTextContent('spam');
  });

  it('check list', () => {
    const {queryByTestId} = renderCheckBoxes(render, 'gems');

    expect(queryByTestId('check-boxes').children).not.toHaveLength(0);
  });

  it('check if snapshot is matching', () => {
    const tree = renderCheckBoxes(renderer.create).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
