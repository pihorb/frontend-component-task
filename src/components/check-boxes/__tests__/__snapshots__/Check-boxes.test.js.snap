// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxes component test check if snapshot is matching  1`] = `
<div
  className="check-boxes"
  data-testid="check-boxes"
>
  <div
    className="check-boxes__title"
  />
  <div
    className="check-boxes__list"
  >
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        create
      </span>
    </div>
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        update
      </span>
    </div>
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        move
      </span>
    </div>
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        delete
      </span>
    </div>
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        view
      </span>
    </div>
    <div
      className="check-box"
      data-testid="check-box"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0.5,
          "pointerEvents": "none",
        }
      }
    >
      <img
        alt="empty"
        className="check-box__img"
        src={Object {}}
      />
      <span
        className="check-box__name"
      >
        share
      </span>
    </div>
  </div>
</div>
`;
